cpp/src/cider/exec/nextgen/transformer/Transformer.cpp

```diff
@@ -44,17 +107,52 @@ static TranslatorPtr generateTranslators(OpPipeline& pipeline) {
   return ptr;
 }
 
-TranslatorPtr Transformer::toTranslator(OpPipeline& pipeline) {
-  // TBD: Currently, we only insert a pair of C2R and R2C at start point and end point of
-  // whole pipeline. Should be designed more properly.
-  auto c2r_node =
-      createOpNode<ColumnToRowNode>(pipeline.front()->getOutputExprs().second);
-  pipeline.insert(pipeline.begin() + 1, c2r_node);
-
-  auto r2c_node =
-      createOpNode<RowToColumnNode>(pipeline.back()->getOutputExprs().second,
-                                    static_cast<ColumnToRowNode*>(c2r_node.get()));
-  pipeline.emplace_back(r2c_node);
+TranslatorPtr Transformer::toTranslator(OpPipeline& pipeline,
+                                        const context::CodegenOptions& co) {
+  CHECK_GT(pipeline.size(), 1);
+  CHECK(isa<QueryFuncInitializer>(pipeline.front()));
+
+  std::vector<PipelineStage> stages;
+  stages.reserve(pipeline.size());
+
+  // Vectorize Project Transformation
+  auto traverse_pivot = ++pipeline.begin();
+  if (co.enable_vectorize && isa<ProjectNode>(*traverse_pivot)) {
+    // Currently, auto-vectorize will be applied to pure project pipeline only.
+    OpNodePtr& curr_op = *traverse_pivot;
+    auto&& [_, exprs] = curr_op->getOutputExprs();
+    ExprPtrVector vectorizable_exprs;
+    vectorizable_exprs.reserve(exprs.size());
+
+    for (auto& expr : exprs) {
+      if (expr->isAutoVectorizable()) {
+        // Move vectorizable exprs out of row-based ProjectNode.
+        vectorizable_exprs.emplace_back(expr);
+        expr.reset();
+      }
+    }
+    exprs.erase(std::remove_if(exprs.begin(),
+                               exprs.end(),
+                               [](ExprPtr& expr) -> bool { return expr == nullptr; }),
+                exprs.end());
+    if (exprs.empty()) {
+      traverse_pivot = pipeline.erase(traverse_pivot);
+    }
+
+    if (!vectorizable_exprs.empty()) {
+      auto vec_proj_iter = pipeline.insert(
+          traverse_pivot, createOpNode<VectorizedProjectNode>(vectorizable_exprs));
+      stages.emplace_back(vec_proj_iter, vec_proj_iter);
+    }
+  }
+
+  if (traverse_pivot != pipeline.end()) {
+    stages.emplace_back(traverse_pivot, --pipeline.end());
+  }
+
+  for (auto&& stage : stages) {
+    stage.generateLoopStage(pipeline);
+  }
```

Comment:
the result for `proj -> agg` will be `c2r -> vec_proj -> r2c -> c2r -> proj -> agg -> r2c`? miss `agg` for the vectorized loop?

Reply:
It should be ```vec_proj->c2r -> (proj) -> agg -> r2c```. ```vec_proj``` has Batch output, so ```generateLoopStage``` will not insert ```C2R``` and ```R2C``` for it.

Reply:
suppose `select sum(a+b), sum(a && b) from t group by c`, `sum(a+b)` will be splited into vec_proj, `sum(a && b), group by c` into `proj -> agg`? then `sum(a+b)` is not right?
otherwise, do we just support nongroupby agg?

Reply:
Not yet, the original pipeline of the SQL likes ```agg```, and transformed to ```c2r->agg->r2c```. Currently, vectorized project only supports pure-project pipeline.

For the given SQL, The ideal solution is that the parser output as ```proj(a+b, a &&b)->agg(sum(a+b), sum(a&&b))```, and  the transformer output as ```vec_proj(a+b, a&&b)->c2r->agg```.

Reply:
good idea